modules/apps/29-fee/ibc_module_test.go

```diff
@@ -468,3 +469,330 @@ func (suite *FeeTestSuite) TestOnChanCloseConfirm() {
+				suite.path.EndpointB.ChannelID,
+				clienttypes.NewHeight(0, 100),
+				0,
+			)
+
+			// set up module and callbacks
+			module, _, err := suite.chainA.App.GetIBCKeeper().PortKeeper.LookupModuleByPort(suite.chainA.GetContext(), ibctesting.TransferPort)
+			suite.Require().NoError(err)
+
+			cbs, ok := suite.chainA.App.GetIBCKeeper().Router.GetRoute(module)
+			suite.Require().True(ok)
+
+			packetId := channeltypes.NewPacketId(suite.path.EndpointA.ChannelID, suite.path.EndpointA.ChannelConfig.PortID, suite.chainA.SenderAccount.GetSequence())
+			identifiedFee := types.NewIdentifiedPacketFee(
+				packetId,
+				types.Fee{
+					ReceiveFee: validCoins,
+					AckFee:     validCoins2,
+					TimeoutFee: validCoins3,
+				},
+				suite.chainA.SenderAccount.GetAddress().String(),
+				[]string{},
+			)
+			err = suite.chainA.GetSimApp().IBCFeeKeeper.EscrowPacketFee(suite.chainA.GetContext(), identifiedFee)
+			suite.Require().NoError(err)
+
+			// malleate test case
+			tc.malleate()
+
+			err = cbs.OnAcknowledgementPacket(suite.chainA.GetContext(), packet, tc.ackBytes, suite.chainB.SenderAccount.GetAddress())
+
+			if tc.expPass {
+				suite.Require().NoError(err, "unexpected error for case: %s", tc.name)
+			}
+
+			if !tc.expPass {
+				suite.Require().Error(err, "%s expected error but returned none", tc.name)
+			}
+		})
+	}
+}
+
+func (suite *FeeTestSuite) TestOnTimeoutPacket() {
+	testCases := []struct {
+		name     string
+		malleate func()
+		expPass  bool
+	}{
+		{
+			"success",
+			func() {},
+			true,
+		},
+		{
+			"no op if identified packet fee doesn't exist",
+			func() {
+				// delete packet fee
+				packetId := channeltypes.NewPacketId(suite.path.EndpointB.ChannelID, suite.path.EndpointB.ChannelConfig.PortID, suite.chainB.SenderAccount.GetSequence())
+				suite.chainA.GetSimApp().IBCFeeKeeper.DeleteFeeInEscrow(suite.chainA.GetContext(), packetId)
+			},
+			true,
+		},
+		{
+			"fee not enabled",
+			func() {
+				suite.chainA.GetSimApp().IBCFeeKeeper.DeleteFeeEnabled(suite.chainA.GetContext(), suite.path.EndpointA.ChannelConfig.PortID, suite.path.EndpointA.ChannelID)
+			},
+			true,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		suite.Run(tc.name, func() {
+			suite.SetupTest()
+
+			// open incentivized channel
+			suite.coordinator.Setup(suite.path)
+
+			// set up coins
+			coin := ibctesting.TestCoin
+			fungibleTokenPacket := transfertypes.NewFungibleTokenPacketData(
+				coin.Denom,
+				sdk.NewInt(100).Uint64(),
+				suite.chainA.SenderAccount.GetAddress().String(),
+				suite.chainB.SenderAccount.GetAddress().String(),
+			)
+			packet := channeltypes.NewPacket(
+				fungibleTokenPacket.GetBytes(),
+				suite.chainA.SenderAccount.GetSequence(),
+				suite.path.EndpointA.ChannelConfig.PortID,
+				suite.path.EndpointA.ChannelID,
+				suite.path.EndpointB.ChannelConfig.PortID,
+				suite.path.EndpointB.ChannelID,
+				clienttypes.NewHeight(0, 100),
+				0,
+			)
```

Comment:
we could optionally make this into a helper function `constructICS20Packet(suite)` since it is used in all the test functions, just a suggestion, not required by any means